pkg/backup/backup_compaction.go

```diff
@@ -0,0 +1,693 @@
+	) {
+		// When a processor is done exporting a span, it will send a progress update
+		// to progCh.
+		for progress := range progCh {
+			var progDetails backuppb.BackupManifest_Progress
+			if err := types.UnmarshalAny(&progress.ProgressDetails, &progDetails); err != nil {
+				log.Errorf(ctx, "unable to unmarshal backup progress details: %+v", err)
+			}
+			for _, file := range progDetails.Files {
+				backupManifest.Files = append(backupManifest.Files, file)
+				backupManifest.EntryCounts.Add(file.EntryCounts)
+			}
+		}
+	}
+
+	store, err := execCtx.ExecCfg().DistSQLSrv.ExternalStorageFromURI(ctx, dest.DefaultURI, execCtx.User())
+	if err != nil {
+		return err
+	}
+	defer store.Close()
+	var tasks []func(context.Context) error
+	tasks = append(tasks, func(ctx context.Context) error {
+		return genSpan(ctx, spanCh)
+	})
+	tasks = append(tasks, func(ctx context.Context) error {
+		return runCompaction(ctx, execCtx, encryption, spanCh, newDetails, backupManifest, progCh, store)
+	})
+	tasks = append(tasks, func(ctx context.Context) error {
+		processProg(ctx, progCh, backupManifest)
+		return nil
+	})
+
+	if err := ctxgroup.GoAndWait(ctx, tasks...); err != nil {
+		return err
+	}
+	return concludeBackupCompaction(ctx, execCtx, store, encryption, kmsEnv, backupManifest)
+}
+
+func runCompaction(
+	ctx context.Context,
+	execCtx sql.JobExecContext,
+	encryption *jobspb.BackupEncryptionOptions,
+	entries chan execinfrapb.RestoreSpanEntry,
+	details jobspb.BackupDetails,
+	manifest *backuppb.BackupManifest,
+	progCh chan execinfrapb.RemoteProducerMetadata_BulkProcessorProgress,
+	store cloud.ExternalStorage,
+) error {
+	defer close(progCh)
+	var encryptionOptions *kvpb.FileEncryptionOptions
+	if encryption != nil {
+		encryptionOptions = &kvpb.FileEncryptionOptions{Key: encryption.Key}
+	}
+	sinkConf := backupsink.SSTSinkConf{
+		ID:        execCtx.ExecCfg().DistSQLSrv.NodeID.SQLInstanceID(),
+		Enc:       encryptionOptions,
+		ProgCh:    progCh,
+		Settings:  &execCtx.ExecCfg().Settings.SV,
+		ElideMode: manifest.ElidedPrefix,
+	}
+	sink, err := backupsink.MakeSSTSinkKeyWriter(sinkConf, store, nil)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := sink.Flush(ctx); err != nil {
+			log.Warningf(ctx, "failed to flush sink: %v", err)
+			logClose(ctx, sink, "SST sink")
+		}
+	}()
+
+	for {
+		select {
+		case entry, ok := <-entries:
+			if !ok {
+				return nil
+			}
+
+			sstIter, err := openSSTs(ctx, execCtx, entry, encryptionOptions, details)
+			if err != nil {
+				return errors.Wrap(err, "opening SSTs")
+			}
+
+			if err := processSpanEntry(ctx, sstIter, sink); err != nil {
+				return errors.Wrap(err, "processing span entry")
+			}
+		}
+	}
+}
+
+func processSpanEntry(
+	ctx context.Context, sstIter mergedSST, sink *backupsink.SSTSinkKeyWriter,
+) error {
+	defer sstIter.cleanup()
+	entry := sstIter.entry
+	trimmedStart, trimmedEnd := storage.MVCCKey{Key: entry.Span.Key}, storage.MVCCKey{Key: entry.Span.EndKey}
```

Comment:
no reason to declare trimmedStart and trimmedEnd up here, right? This is slightly confusing as they are not trimmed. Seems simpler to one line it:
```
trimmedStart := storage.MVCCKey{ Key: bytes.TrimPrefix(entry.Span.Key, prefix)}
// same for endkey
```

Reply:
Oh good point